### PR TITLE
feat: Respect custom `timestamp` logger option

### LINKED ISSUE #63: Logger does not accept provided timestamp logger option
## Description

The logger does not use the provided `timestamp` logger option when creating the logger instance.
The `timestamp` is always [clobbered](https://github.com/seek-oss/logger/blob/e5378439bf6a8a21ac92ea9a24943460767b8734/src/index.ts#L40)

E.g.
```typescript
import createLogger from "@seek/logger";

const loggerWithTimestampOption = createLogger({
  timestamp: false
});

logger.info({
  timestamp: 'actual event time'
})
```

expect to get
```
{"level":30,"timestamp":"actual event time"}
```
but get
```
{"level":30,"timestamp":"2023-01-03T02:34:21.370Z","timestamp":"actual event time"}
```

## Possible Implementation

Use provided `timestamp` logger option if it is not undefined, otherwise use the default timestamp function.

## Context

We have a [worker](https://github.com/SEEK-Jobs/hirer-audit-cloudwatch-publisher) that receives audit log events through SNS, then we use `@seek/logger` to log received events to CloudWatch and forward them to Splunk.

The `timestamp` in the log could be confusing since it is the time that the worker logs events to CloudWatch and
we would like to record the actual time for the event(The audit logs have their own `eventDateUtc`).

### PR DESCRIPTION
**Pull request by JasonZzy0528**
## Purpose

I would like to use provided `timestamp` logger option when creating the logger instance.
Currently, the input `timestamp` logger option is overridden by [the default timestamp function](https://github.com/seek-oss/logger/blob/master/src/index.ts#L40)

## Approach

Use provided `timestamp` logger option if it is not undefined, otherwise use the default timestamp function.

## Notes

This can be a breaking change since several places(e.g. [hirer-contracts-api](https://github.com/SEEK-Jobs/hirer-contracts-api/blob/master/src/util/logger.ts#L16), [hirer-budgets-api](https://github.com/SEEK-Jobs/hirer-budgets-api/blob/master/src/util/logging.ts#L45), and [etc.](https://github.com/search?q=org%3ASEEK-Jobs+%22%40seek%2Flogger%22+createLogger+timestamp&type=Code)) are using the timestamp logger option and this change may generate different timestamp field.

## Issues

Closes #63

### CODE REVIEW

File: src/index.test.ts [hunk 1]
```diff
@@ -455,3 +455,54 @@ test('enforces a specified object interface', async () => {
...
+      timestamp: false,
+    },
+    stream,
+  );
+
+  logger.info<ExampleMessageContext>(
+    {
+      activity: 'Testing Logger',
+      // @ts-expect-error
+      propertyNotAllowed: 'Linting error',
```
Comment (72636c) [inline src/index.test.ts:472]: This does not seem relevant to the test case:

```suggestion
```

File: src/index.test.ts [hunk 2]
```diff
@@ -455,3 +455,54 @@ test('enforces a specified object interface', async () => {
...
+        foo: 0xf00,
+      },
+    },
+    'Test log entry',
+  );
+  const log: any = await once(stream, 'data');
+  expect(log).not.toHaveProperty('timestamp');
+});
+
+test('should log customized timestamp if timestamp logger option is not false or undefined', async () => {
```
Comment (72636c) [inline src/index.test.ts:483]:
```suggestion
test('should log customized timestamp if timestamp logger option is supplied', async () => {
```

File: src/index.test.ts [hunk 3]
```diff
@@ -455,3 +455,54 @@ test('enforces a specified object interface', async () => {
...
+      timestamp: () => `,"timestamp":"${mockTimestamp}"`,
+    },
+    stream,
+  );
+
+  logger.info<ExampleMessageContext>(
+    {
+      activity: 'Testing Logger',
+      // @ts-expect-error
+      propertyNotAllowed: 'Linting error',
```
Comment (72636c) [inline src/index.test.ts:498]:
```suggestion
```

File: src/index.ts [hunk 4]
```diff
@@ -37,7 +37,10 @@ export default (
     ...formatters,
     ...opts.formatters,
   };
-  opts.timestamp = () => `,"timestamp":"${new Date().toISOString()}"`;
+
+  if (opts.timestamp === undefined) {
+    opts.timestamp = () => `,"timestamp":"${new Date().toISOString()}"`;
+  }
```
Comment (72636c) [inline src/index.ts:43]:
```suggestion
  opts.timestamp ??= () => `,"timestamp":"${new Date().toISOString()}"`;
```

**Review by 72636c — APPROVED** [review]

File: src/index.test.ts (hunk 1 above)
Comment (JasonZzy0528) [reply src/index.test.ts]: Good catch! @72636c

**Comment by 72636c** [comment]
Thanks a bunch for the detailed description and fix @JasonZzy0528!
